File: LeetCode/C/142-linked-list-环形链表-ii.cpp
/*
 * @lc app=leetcode.cn id=142 lang=cpp
 *
 * [142] 环形链表 II
 */
#include "main.h"

// @lc code=start
class Solution {
  public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *p = head;
        unordered_set<ListNode *> s;
        while (p) {
            if (s.insert(p).second == false)
                return p;
            p = p->next;
        }
        return nullptr;
    }
};
// @lc code=end

File: LeetCode/Java/142-linked-list-环形链表-ii.java
import java.util.*;

/*
 * @lc app=leetcode.cn id=142 lang=java
 *
 * [142] 环形链表 II
 */

// @lc code=start
class Solution {
    public ListNode detectCycle(ListNode head) {
        Set<ListNode> set = new HashSet<>();
        ListNode p = head;
        while (p != null) {
            if (set.contains(p)) {
                return p;
            }
            set.add(p);
            p = p.next;
        }
        return null;
    }
}
// @lc code=end
